pkg/events/event_ringbuffer.go

```diff
@@ -0,0 +1,220 @@
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package events
+
+import (
+	"math"
+	"sync"
+
+	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
+)
+
+type eventRange struct {
+	start uint64
+	end   uint64
+}
+
+// eventRingBuffer A specialized circular buffer to store event objects.
+//
+// Unlike to regular circular buffers, existing entries are never directly removed and new entries can be added if the buffer is full.
+// In this case, the oldest entry is overwritten and can be collected by the GC.
+// Each event has an ID, however, this mapping is not stored directly. If needed, we calculate the id
+// of the event based on slice positions.
+//
+// Retrieving the records can be achieved with GetEventsFromID and GetRecentEntries.
+type eventRingBuffer struct {
+	events   []*si.EventRecord
+	capacity uint64 // capacity of the buffer, never changes
+	head     uint64 // position of the next element (no tail since we don't remove elements)
+	full     bool   // indicates whether the buffer if full - once it is, it stays full
+	id       uint64 // unique id of an event record
+
+	sync.RWMutex
+}
+
+// Add adds an event to the ring buffer. If the buffer is full, the oldest element is overwritten.
+// This method never fails.
+func (e *eventRingBuffer) Add(event *si.EventRecord) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.events[e.head] = event
+	if !e.full {
+		e.full = e.head == e.capacity-1
+	}
+	e.head = (e.head + 1) % e.capacity
+	e.id++
+}
+
+func (e *eventRingBuffer) GetEventsFromID(id uint64, count uint64) ([]*si.EventRecord, uint64) {
+	e.RLock()
+	defer e.RUnlock()
+	lowest := e.getLowestId()
+
+	pos, idFound := e.id2pos(id)
+	if !idFound {
+		return nil, lowest
+	}
+
+	if e.full && pos > e.head {
+		r1 := &eventRange{
+			start: pos,
+			end:   e.capacity,
+		}
+		r2 := &eventRange{
+			start: 0,
+			end:   e.head,
+		}
+		return e.getEntriesFromRanges(r1, r2, count), lowest
+	}
+
+	end := e.head
+	if e.full && e.head == 0 {
+		// Special case: buffer is full and head is pointing at the beginning.
+		// Need to set explicitly to get [0..capacity] range because e.head
+		// never points to e.capacity.
+		end = e.capacity
+	}
+	return e.getEntriesFromRanges(&eventRange{
+		start: pos,
+		end:   end,
+	}, nil, count), lowest
```

Comment:
We know the count here why are we not enforcing the count here instead of passing this as a limitation to be applied on top of the ranges we have created?
```
func (e *eventRingBuffer) GetEventsFromID(id uint64, count uint64) ([]*si.EventRecord, uint64) {
	e.RLock()
	defer e.RUnlock()
	lowest := e.getLowestId()

	pos, idFound := e.id2pos(id)
	if !idFound {
		return nil, lowest
	}
	// limit count to the capacity
	count = min(count, e.capacity)
	// possible wrap case full buffer and pos after or on the current head
	if e.full && pos >= e.head {
		// first range never passes the end of the slice
		end := min(pos+count, e.capacity)
		r1 := &eventRange{
			start: pos,
			end:   end,
		}
		// second range only if still events left to fetch
		var r2 *eventRange
		end = pos + count - e.capacity
		if end > 0 {
			// never fetch pass the current head
			end = min(end, e.head)
			r2 = &eventRange{
				start: 0,
				end:   end,
			}
		}
		return e.getEntriesFromR(r1, r2), lowest
	}

	// no wrapping: either limited by head position or by the count
	end := min(pos+count, e.head)
	return e.getEntriesFromR(&eventRange{
		start: pos,
		end:   end,
	}, nil), lowest
}

// min a utility function to return the smallest value of two unsigned int
func min(a, b uint64) uint64 {
	m := a
	if b < a {
		m = b
	}
	return m
}
```

Reply:
`getEntriesFromRange()` can then be as simple as it was a couple of days ago without all the count handling and ignoring the second range if limited by count